news/scripts/news_fetcher.py

```diff
@@ -0,0 +1,209 @@
+parser.add_argument('-m', '--max_req',
+                    help='Maximum number of requests to be sent to StockNewsAPI',
+                    type=int,
+                    default=MAX_API_PAGE_REQUEST)
+
+
+class Fetcher:
+    """
+    Exhaustive stock news articles fetcher for StockNewsAPI
+    """
+
+    def __init__(self, start_date, end_date, num_items, max_req):
+        self.tickers: List[str] = []
+        self.request_url: str = STOCK_NEWS_API_URL
+        self.num_items: int = num_items
+        self.tot_pgs_api: int = -1
+        self.cur_page: int = 1
+        self.start_date: str = start_date
+        self.end_date: str = end_date
+        self.max_req: int = max_req
+        self.count = 0
+        self.api_key = API_KEY
+
+    def _prepare_tickers(self):
+        """
+        Fetches instrument tickers for which news articles are needed
+        """
+        instr_req: requests.Response = requests.get('{}instruments/'.format(GRAPHELIER_SERVICE_URL))
+        try:
+            instr_req.raise_for_status()
+        except requests.exceptions.HTTPError as e:
+            print('Could not retrieve instruments from graphelier service: {}'.format(str(e)))
+            sys.exit(1)
+        tickers: List[str] = instr_req.json()
+        self.tickers = tickers
+
+    def _add_query_param(self, key: str, value: str):
+        """
+        Adds single query param to request url to StockNewsAPI
+        """
+        if not self.request_url.endswith('?'):
+            self.request_url += '&'
+        self.request_url += '{}={}'.format(key, value)
+
+    def _prepare_query_string(self):
+        """
+        Builds request string, with format string on the page number
+        """
+        self._add_query_param('items', str(self.num_items))
+        self._add_query_param('type', 'article')
+        self._add_query_param('tickers', ','.join(self.tickers))
+        self._add_query_param('date', '{}-{}'.format(self.start_date, self.end_date))
+        self._add_query_param('token', self.api_key)
+        self._add_query_param('page', '{}')
+
+    def run(self):
+        """
+        Execute requests and write to db for every page
+        """
+        self._prepare_tickers()
+        self._prepare_query_string()
+
+        # while self.cur_page <= 1:
+        while self.cur_page <= self.tot_pgs_api or self.tot_pgs_api == -1:
+            pg_req: requests.Response = requests.get(self.request_url.format(self.cur_page))
+            try:
+                pg_req.raise_for_status()
+            except requests.exceptions.HTTPError as e:
+                print('Error retrieving news from StockNewsAPI: {}'.format(str(e)))
+                return
+            resp_body = pg_req.json()
+            if 'data' not in resp_body or len(resp_body['data']) == 0:
+                # additional error check
+                print('No articles found for specified tickers')
+                break
+            if 'total_pages' in resp_body and self.tot_pgs_api == -1:
```

Comment:
what does this do?

Reply:
sets the max page number that's given in the first response by the API